File: lib/src/main/java/com/orientsec/easysocket/session/BlockingReader.java
package com.orientsec.easysocket.session;

import android.net.TrafficStats;

import com.orientsec.easysocket.EasyExecutor;
import com.orientsec.easysocket.HeadParser;
import com.orientsec.easysocket.Options;
import com.orientsec.easysocket.Packet;
import com.orientsec.easysocket.client.BaseSocketClient;
import com.orientsec.easysocket.error.EasyException;
import com.orientsec.easysocket.error.ErrorCode;
import com.orientsec.easysocket.error.ErrorType;

import java.io.IOException;
import java.io.InputStream;
import java.net.Socket;

/**
 * A class responsible for reading data from a socket in a blocking manner.
 * This class extends the `Looper` class and implements the `Reader` interface.
 */
public class BlockingReader extends Looper implements Reader {
    // Input stream for reading data from the socket
    private InputStream inputStream;

    // Parser for handling the packet headers
    private final HeadParser headParser;

    // Session associated with this reader
    private final OperableSession session;

    // Socket used for communication
    private final Socket socket;

    // Configuration options for the reader
    private final Options options;

    // EasyExecutor for executing tasks
    private final EasyExecutor mainExecutor;

    /**
     * Constructs a BlockingReader instance with the specified parameters.
     *
     * @param session The session associated with this reader.
     * @param socket  The socket used for communication.
     * @param client  The BaseSocketClient instance providing options and executors.
     */
    BlockingReader(OperableSession session, Socket socket, BaseSocketClient client) {
        super(session.getLogger());
        this.session = session;
        this.socket = socket;
        this.mainExecutor = client.getMainExecutor();
        this.options = client.getOptions();
        this.headParser = client.getHeadParser();
    }

    /**
     * Reads data from the input stream, parses the packet, and handles it.
     *
     * @throws Exception If an error occurs during reading or parsing.
     */
    @Override
    public void read() throws Exception {
        int headLength = headParser.headSize();
        byte[] headBytes = new byte[headLength];
        readInputStream(inputStream, headBytes);
        HeadParser.Head head = headParser.parseHead(headBytes);
        int bodyLength = head.getPacketSize();
        if (bodyLength > options.getMaxReadSizeInKB() * 1024) {
            throw new Exception("packet size too large: " + bodyLength);
        } else if (bodyLength >= 0) {
            byte[] data = new byte[bodyLength];
            readInputStream(inputStream, data);
            Packet packet = headParser.decodePacket(head, data);
            session.handlePacket(packet);
        } else {
            throw new Exception("negative packet size : " + bodyLength);
        }
    }

    /**
     * Reads data from the input stream into the provided byte array.
     *
     * @param inputStream The input stream to read from.
     * @param data        The byte array to store the read data.
     * @throws IOException If an error occurs during reading.
     */
    private void readInputStream(InputStream inputStream, byte[] data) throws IOException {
        int readCount = 0; // Number of bytes successfully read
        int count = data.length;
        while (readCount < count) {
            int len = inputStream.read(data, readCount, count - readCount);
            if (len == -1) {
                throw new IOException("input stream closed");
            }
            readCount += len;
        }
    }

    /**
     * Prepares the reader before starting the loop.
     *
     * @throws IOException If an error occurs while setting up the input stream.
     */
    @Override
    protected void beforeLoop() throws IOException {
        TrafficStats.setThreadStatsTag(options.getReadStatsTag());
        inputStream = socket.getInputStream();
    }

    /**
     * Executes the reading logic in the loop thread.
     *
     * @throws Exception If an error occurs during reading.
     */
    @Override
    protected void runInLoopThread() throws Exception {
        read();
    }

    /**
     * Cleans up resources and handles errors when the loop finishes.
     */
    @Override
    protected synchronized void loopFinish() {
        TrafficStats.clearThreadStatsTag();
        EasyException e = EasyException.create(ErrorCode.READ_EXIT, ErrorType.CONNECT,
                "socket read aborted", session.getSuffix(), error);
        if (isRunning()) {
            mainExecutor.execute(() -> session.close(e));
        }
    }
}

File: lib/src/main/kotlin/com/orientsec/easysocket/session/BlockingReader.kt
package com.orientsec.easysocket.session

import android.net.TrafficStats
import com.orientsec.easysocket.HeadParser
import com.orientsec.easysocket.Options
import com.orientsec.easysocket.client.BaseSocketClient
import com.orientsec.easysocket.error.EasyException
import com.orientsec.easysocket.error.ErrorCode
import com.orientsec.easysocket.error.ErrorType
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.IOException
import java.io.InputStream
import java.net.Socket

/**
 * A class responsible for reading data from a socket in a blocking manner.
 * This class extends the `Looper` class and implements the `Reader` interface.
 */
class BlockingReader(
    private val session: OperableSession,
    private val socket: Socket,
    client: BaseSocketClient
) : Looper(session.logger), Reader {

    // Input stream for reading data from the socket
    private lateinit var inputStream: InputStream

    // Parser for handling the packet headers
    private val headParser: HeadParser = client.headParser

    // Maximum size of the packet to read, in bytes
    private val maxReadSize: Long = client.options.maxReadSizeKb * 1024L

    // Configuration options for the reader
    private val options: Options = client.options

    private val scope = client.scope

    /**
     * Reads data from the input stream, parses the packet, and handles it.
     */
    override suspend fun read() {
        val headLength = headParser.headSize()
        val headBytes = ByteArray(headLength)
        readInputStream(inputStream, headBytes)
        val head = headParser.parseHead(headBytes)
        val packetSize = head.packetSize
        if (packetSize > maxReadSize) {
            throw Exception(
                "packet size: $packetSize is large than max size: $maxReadSize"
            )
        } else if (packetSize >= 0) {
            val data = ByteArray(packetSize)
            readInputStream(inputStream, data)
            val packet = headParser.decodePacket(head, data)
            session.handlePacket(packet)
        } else {
            throw Exception("negative packet size: $packetSize")
        }
    }

    /**
     * Reads data from the input stream into the provided byte array.
     */
    private suspend fun readInputStream(inputStream: InputStream, data: ByteArray) {
        withContext(Dispatchers.IO) {
            var readCount = 0
            val count = data.size
            while (readCount < count) {
                val len = inputStream.read(data, readCount, count - readCount)
                if (len == -1) {
                    throw IOException("input stream closed")
                }
                readCount += len
            }
        }
    }

    override fun beforeLoop() {
        TrafficStats.setThreadStatsTag(options.readStatsTag)
        inputStream = socket.getInputStream()
    }

    override suspend fun runInLoopThread() {
        read()
    }

    @Synchronized
    override fun loopFinish() {
        TrafficStats.clearThreadStatsTag()
        val e = EasyException(
            ErrorCode.READ_EXIT, ErrorType.CONNECT,
            "socket read aborted", session.suffix, error
        )
        if (isRunning()) {
            scope.launch { session.close(e) }
        }
    }
}
